File: apps/web/src/ai/__tests__/planned-structure.test.ts
// Golden-fixture tests for the defensive `planned_workouts.structure` readers
// (src/ai/planned-structure.ts) and the KTD6 intensity display formatter
// (src/components/planned/planned-workout-view.ts).
//
// Both this file and daily-athlete/test/models/planned_structure_test.dart
// (a parallel unit, U2) load the SAME rows from
// packages/shared/test-fixtures/planned-structure-vectors.json so the
// TypeScript and Dart readers/formatters cannot silently disagree (KTD2a).
// Do not hand-write separate expectations here -- assert against the shared
// fixture's fields directly.

import { readFileSync } from "node:fs";
import path from "node:path";
import { describe, expect, it } from "vitest";

import {
  readStructureDurationSeconds,
  readStructureIntensityTarget,
  readStructureLoad,
} from "@/ai/planned-structure";
import {
  extractLegacySteps,
  formatIntensityTarget,
} from "@/components/planned/planned-workout-view";

const FIXTURE_PATH = path.resolve(
  __dirname,
  "../../../../../packages/shared/test-fixtures/planned-structure-vectors.json"
);

interface DurationLoadIntensityRow {
  name: string;
  structure_input: Record<string, unknown>;
  planned_load_column: number | null;
  expected_duration_s: number | null;
  expected_load: number | null;
  expected_intensity_target: { kind: string; value: number } | null;
  expected_display_string: string | null;
}

interface LegacyStepExpectation {
  label: string | null;
  duration_s: number | null;
  display_string: string | null;
}

interface LegacyStepsRow {
  name: string;
  structure_input: Record<string, unknown>;
  expected_steps: LegacyStepExpectation[];
}

interface Fixture {
  duration_load_intensity: DurationLoadIntensityRow[];
  legacy_steps: LegacyStepsRow[];
}

const fixture = JSON.parse(readFileSync(FIXTURE_PATH, "utf-8")) as Fixture;

describe("planned-structure readers (KTD2a golden fixture)", () => {
  for (const row of fixture.duration_load_intensity) {
    it(row.name, () => {
      expect(readStructureDurationSeconds(row.structure_input)).toBe(row.expected_duration_s);
      expect(readStructureLoad(row.structure_input, row.planned_load_column)).toBe(
        row.expected_load
      );
      expect(readStructureIntensityTarget(row.structure_input)).toEqual(
        row.expected_intensity_target
      );

      const intensityTarget = readStructureIntensityTarget(row.structure_input);
      expect(formatIntensityTarget(intensityTarget)).toBe(row.expected_display_string);
    });
  }
});

describe("legacy step extraction (KTD5 golden fixture)", () => {
  for (const row of fixture.legacy_steps) {
    it(row.name, () => {
      const steps = extractLegacySteps(row.structure_input);
      expect(steps).toEqual(row.expected_steps);
    });
  }
});

File: daily-athlete/test/models/planned_structure_test.dart
// Loads the shared golden-fixture vectors from
// packages/shared/test-fixtures/planned-structure-vectors.json and asserts
// planned_structure.dart's readers/formatter/step-extractor against them.
// The same rows are asserted against by
// apps/web/src/ai/__tests__/planned-structure.test.ts (KTD2a in
// docs/plans/2026-08-27-001-feat-planned-workout-detail-rendering-plan.md),
// so a semantic disagreement between the Dart port and the TS original fails
// here rather than passing silently on one side.

import 'dart:convert';
import 'dart:io';

import 'package:daily_athlete/models/planned_structure.dart';
import 'package:daily_athlete/models/workout_report.dart' show IntensityTargetKind;
import 'package:flutter_test/flutter_test.dart';

/// Walks up from the current working directory to find the repo root (the
/// directory containing `packages/shared/test-fixtures/...`), so the test
/// resolves the fixture regardless of whether `flutter test` is invoked from
/// `daily-athlete/` (the documented convention) or the monorepo root.
File _locateFixture() {
  const relative = 'packages/shared/test-fixtures/planned-structure-vectors.json';
  var dir = Directory.current;
  for (var i = 0; i < 6; i++) {
    final candidate = File('${dir.path}/$relative');
    if (candidate.existsSync()) return candidate;
    final parent = dir.parent;
    if (parent.path == dir.path) break;
    dir = parent;
  }
  throw FileSystemException(
    'Could not locate planned-structure-vectors.json above ${Directory.current.path}',
  );
}

Map<String, dynamic>? _asMap(dynamic v) => v == null ? null : Map<String, dynamic>.from(v as Map);

void main() {
  final fixtureFile = _locateFixture();
  final fixture = jsonDecode(fixtureFile.readAsStringSync()) as Map<String, dynamic>;

  group('duration_load_intensity vectors', () {
    final rows = fixture['duration_load_intensity'] as List;
    for (final row in rows) {
      final r = row as Map<String, dynamic>;
      test(r['name'] as String, () {
        final structure = _asMap(r['structure_input']);
        final plannedLoadColumn = r['planned_load_column'] as num?;

        expect(readStructureDurationSeconds(structure), r['expected_duration_s']);
        expect(readStructureLoad(structure, plannedLoadColumn), r['expected_load']);

        final expectedIntensity = _asMap(r['expected_intensity_target']);
        final intensity = readStructureIntensityTarget(structure);
        if (expectedIntensity == null) {
          expect(intensity, isNull);
        } else {
          expect(intensity, isNotNull);
          final expectedKind = switch (expectedIntensity['kind'] as String) {
            'ftp_pct' => IntensityTargetKind.ftpPct,
            'zone' => IntensityTargetKind.zone,
            'pace_s_per_km' => IntensityTargetKind.paceSPerKm,
            final other => throw ArgumentError('Unknown kind: $other'),
          };
          expect(intensity!.kind, expectedKind);
          expect(intensity.value, expectedIntensity['value']);
        }

        final expectedDisplay = r['expected_display_string'] as String?;
        if (expectedDisplay == null) {
          expect(intensity, isNull, reason: 'no display string expected implies no target');
        } else {
          expect(formatIntensityTarget(intensity!), expectedDisplay);
        }
      });
    }
  });

  group('legacy_steps vectors', () {
    final rows = fixture['legacy_steps'] as List;
    for (final row in rows) {
      final r = row as Map<String, dynamic>;
      test(r['name'] as String, () {
        final structure = _asMap(r['structure_input']);
        final expectedSteps = r['expected_steps'] as List;

        final steps = extractPlannedSteps(structure);

        expect(steps.length, expectedSteps.length);
        for (var i = 0; i < expectedSteps.length; i++) {
          final expected = expectedSteps[i] as Map<String, dynamic>;
          final actual = steps[i];
          expect(actual.label, expected['label']);
          expect(actual.durationS, expected['duration_s']);
          expect(actual.displayString, expected['display_string']);
        }
      });
    }
  });
}
